### DISCUSSION TITLE
jsPsych.pluginAPI.compareKeys

### DISCUSSION THREAD

**Discussion by limkhaiin** [jspsych/jsPsych#2114] [General]
hello,

I tried to use this function to compare the response and correct response, but it always returns 'false' even though both numeric response are the same. I am using 'audio-button-response' and the data return 1 or 0. Here is my code:

```js
var trial = {
  timeline_variables: trial_parameters,
  randomize_order: true,
  timeline: [{
    type: 'audio-button-response',
    stimulus: jsPsych.timelineVariable('audio'),
    choices: jsPsych.timelineVariable('images'),
    button_html: '<img style="width:100px;" onmouseout="this.src=\'img/%choice%.png\'" src="img/%choice%.png" />',
    margin_horizontal: "20px",
    on_finish: function(data){
        data.correct = jsPsych.pluginAPI.compareKeys(data.response,data.correct_response)}
  }]
}

timeline.push(trial);
```

In my trial_parameters, there are 'audio', 'images', 'correct_response' keys.

Thanks,

**Comment by jodeleeuw** [comment]
Hi @limkhaiin ,

The parameters in `timeline_variables` are not automatically added to the `data` for the trial. You can add them like this:

```js
var trial = {
  timeline_variables: trial_parameters,
  randomize_order: true,
  timeline: [{
    type: 'audio-button-response',
    stimulus: jsPsych.timelineVariable('audio'),
    choices: jsPsych.timelineVariable('images'),
    button_html: '<img style="width:100px;" onmouseout="this.src=\'img/%choice%.png\'" src="img/%choice%.png" />',
    margin_horizontal: "20px",
    data: {
      correct_response: jsPsych.timelineVariable('correct_response')
    },
    on_finish: function(data){
        data.correct = jsPsych.pluginAPI.compareKeys(data.response,data.correct_response)}
  }]
}

**Reply by limkhaiin** [reply]
Thanks, but I still don't get correct answer with `data.correct`. I think just as Gilbert said, we should avoid using `compareKeys()`.

**Comment by becky-gilbert** [comment]
Hi @limkhaiin, Josh's answer should fix the problem, but I also wanted to note that the `jsPsych.pluginAPI.compareKeys` function is really just intended for comparing key press values, which are strings, like 'a', ' ', etc. Whereas you're comparing button response values, which are numbers. Right now the `compareKeys` function works with numbers, but that might change in future jsPsych releases, so I recommend using a normal equality test:
```js
on_finish: function(data){
  data.correct = data.response === data.correct_response;
}
```
Lots of other people have been using the `compareKeys` function for button responses as well, so we should probably clarify this and check for errors in the jsPsych documentation!

**Reply by limkhaiin** [reply]
Thanks, it works ! And note that the boolean expression should be `'=='` just for reference !

**Reply by becky-gilbert** [reply]
Hmm, the [strict equality operator](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Strict_equality) (`===`) will only be true if the two variables have the same type, so it should work if `data.response` and `data.correct_response` are both numbers. Whereas the regular [equality operator](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Equality) (`==`) will attempt to convert the two variables to the same type before comparing them. So if your code is only working with the regular equality operator, then that suggests that the `response` and `correct_response` values have different types (e.g. number and string).

Anyway, glad it's working!
